pkg/controller/migcluster/validation.go

```diff
@@ -3,38 +3,208 @@ package migcluster
 // Returns error and the total error conditions set.
-func (r ReconcileMigCluster) validate(assetCollection *migapi.MigCluster) (error, int) {
+func (r ReconcileMigCluster) validate(cluster *migapi.MigCluster) (error, int) {
 	totalSet := 0
 	var err error
+	nSet := 0
 
-	// Apply changes
-	err = r.Update(context.TODO(), assetCollection)
+	// registry cluster
+	err, nSet = r.validateRegistryCluster(cluster)
+	if err != nil {
+		return err, 0
+	}
+	totalSet += nSet
+
+	// SA secret
+	err, nSet = r.validateSaSecret(cluster)
+	if err != nil {
+		return err, 0
+	}
+	totalSet += nSet
+
+	// Ready
+	cluster.Status.SetReady(totalSet == 0, ReadyMessage)
+
+	// Apply changes.
+	err = r.Update(context.TODO(), cluster)
 	if err != nil {
 		return err, 0
 	}
 
 	return err, totalSet
 }
+
+func (r ReconcileMigCluster) validateRegistryCluster(cluster *migapi.MigCluster) (error, int) {
+	ref := cluster.Spec.ClusterRef
+
+	// NotSet
+	if !migref.RefSet(ref) {
+		cluster.Status.SetCondition(migapi.Condition{
+			Type:    InvalidClusterRef,
+			Status:  True,
+			Reason:  NotSet,
+			Message: InvalidClusterRefMessage,
+		})
+		return nil, 1
+	}
+
+	err, storage := r.getCluster(ref)
+	if err != nil {
+		return err, 0
+	}
+
+	// NotFound
+	if storage == nil {
+		cluster.Status.SetCondition(migapi.Condition{
+			Type:    InvalidClusterRef,
+			Status:  True,
+			Reason:  NotFound,
+			Message: InvalidClusterRefMessage,
+		})
+		return nil, 1
+	} else {
+		cluster.Status.DeleteCondition(InvalidClusterRef)
+	}
+
+	return nil, 0
+}
+
+func (r ReconcileMigCluster) getCluster(ref *kapi.ObjectReference) (error, *crapi.Cluster) {
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+
+	cluster := crapi.Cluster{}
+	err := r.Get(context.TODO(), key, &cluster)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		} else {
+			return err, nil
+		}
+	}
+
+	return nil, &cluster
+}
+
+func (r ReconcileMigCluster) validateSaSecret(cluster *migapi.MigCluster) (error, int) {
+	ref := cluster.Spec.ServiceAccountSecretRef
+
+	if cluster.Spec.IsHostCluster {
+		cluster.Status.DeleteCondition(InvalidSaSecretRef)
+		cluster.Status.DeleteCondition(InvalidSaToken)
+		return nil, 0
+	}
+
+	// NotSet
+	if !migref.RefSet(ref) {
+		cluster.Status.SetCondition(migapi.Condition{
+			Type:    InvalidSaSecretRef,
+			Status:  True,
+			Reason:  NotSet,
+			Message: InvalidSaSecretRefMessage,
+		})
+		cluster.Status.DeleteCondition(InvalidSaToken)
+		return nil, 1
+	}
+
+	err, secret := r.getSecret(ref)
+	if err != nil {
+		return err, 0
+	}
+
+	// NotFound
+	if secret == nil {
+		cluster.Status.SetCondition(migapi.Condition{
+			Type:    InvalidSaSecretRef,
+			Status:  True,
+			Reason:  NotFound,
+			Message: InvalidSaSecretRefMessage,
+		})
+		cluster.Status.DeleteCondition(InvalidSaToken)
+		return nil, 1
+	} else {
+		cluster.Status.DeleteCondition(InvalidSaSecretRef)
+	}
+
+	// saToken
+	token, found := secret.Data["saToken"]
```

Comment:
I wonder if we should put this "magic string" in a var somewhere since we have it in multiple places now. +1 for checking if the data exists

Reply:
Yeah a shared _const_ is a very good idea.